src/MediatR/Internal/ObjectTypeEqualityComparer.cs

```diff
@@ -0,0 +1,10 @@
+﻿using System.Collections.Generic;
+
+namespace MediatR.Internal;
+
+public class ObjectTypeEqualityComparer : IEqualityComparer<object?>
```

Comment:
The class `Type` already implement `Equals` and `GetHashCode` so I don't think this is necessary.
Also if you look at its `Type.Equals` implementation it does `object.ReferenceEquals(this.UnderlyingSystemType, o.UnderlyingSystemType)` which seems better.

Reply:
This scenario then will result in duplication:
```csharp
class Foo {}
class Bar : Foo {}

services.AddTransient<INotificationHandler<Foo>, FooHandler>();
services.AddTransient<INotificationHandler<Bar>, FooHandler>();

mediator.Publish(new Bar());
```